src/serac/numerics/equation_solver.cpp

```diff
@@ -361,64 +438,93 @@ class TrustRegion : public mfem::NewtonSolver {
     }
   }
 
+  /// compute the energy of the linearized system for a given solution vector z
+  template <typename HessVecFunc>
+  double computeEnergy(const mfem::Vector& r_local, const HessVecFunc& H, const mfem::Vector& z) const
+  {
+    double       rz = Dot(r_local, z);
+    mfem::Vector tmp(r_local);
+    tmp *= 0.0;
+    H(z, tmp);
+    return rz + 0.5 * Dot(z, tmp);
+  }
+
   /// Minimize quadratic sub-problem given residual vector, the action of the stiffness and a preconditioner
   template <typename HessVecFunc, typename PrecondFunc>
   void solveTrustRegionModelProblem(const mfem::Vector& r0, mfem::Vector& rCurrent, HessVecFunc hess_vec_func,
                                     PrecondFunc precond, const TrustRegionSettings& settings, double& trSize,
                                     TrustRegionResults& results) const
   {
     SERAC_MARK_FUNCTION;
-    // minimize r@z + 0.5*z@J@z
+    // minimize r0@z + 0.5*z@J@z
     results.interior_status     = TrustRegionResults::Status::Interior;
     results.cg_iterations_count = 0;
 
     auto& z      = results.z;
     auto& cgIter = results.cg_iterations_count;
     auto& d      = results.d;
     auto& Pr     = results.Pr;
-    auto& Hd     = results.Hd;
+    auto& Hd     = results.H_d;
 
     const double cg_tol_squared = settings.cg_tol * settings.cg_tol;
 
-    if (Dot(r0, r0) <= cg_tol_squared && cgIter >= settings.min_cg_iterations) {
+    if (Dot(r0, r0) <= cg_tol_squared && settings.min_cg_iterations == 0) {
       return;
     }
 
     rCurrent = r0;
     precond(rCurrent, Pr);
-    d = 0.0;
-    add(d, -1.0, Pr, d);  // d = -Pr
+
+    // d = -Pr
+    d = Pr;
+    d *= -1.0;
 
     z          = 0.0;
     double zz  = 0.;
     double rPr = Dot(rCurrent, Pr);
     double zd  = 0.0;
     double dd  = Dot(d, d);
 
+    // std::cout << "initial energy = " << computeEnergy(r0, hess_vec_func, z) << std::endl;
+
     for (cgIter = 1; cgIter <= settings.max_cg_iterations; ++cgIter) {
+      // check if this is a decent direction
+      if (Dot(d, rCurrent) > 0) {
+        d *= -1;
+        results.interior_status = TrustRegionResults::Status::NonDescentDirection;
+      }
+
       hess_vec_func(d, Hd);
       const double curvature = Dot(d, Hd);
       const double alphaCg   = curvature != 0.0 ? rPr / curvature : 0.0;
```

Comment:
Ok, this prevents getting infinity or nan, but setting `alphaCg` in that case to zero means that the solver will stagnate at this point. Should we set it to 1, and then the "go to the boundary" step at least tries to move to a better point?

Reply:
following the logic down a few lines, this is indeed what actually happens already.  the alphaCg is used for a trial solution.  If the curvature is <= 0, we then do the project to boundary stuff.